tests/integration/test_node_port.py

```diff
@@ -0,0 +1,156 @@
+            application_name=DATA_INTEGRATOR,
+            series="jammy",
+            config={"database-name": "test"},
+            num_units=1,
+        ),
+    )
+
+    mysql_app, application_app = applications[0], applications[2]
+
+    async with ops_test.fast_forward():
+        logger.info("Waiting for mysqlrouter to be in BlockedStatus")
+        await ops_test.model.block_until(
+            lambda: ops_test.model.applications[MYSQL_ROUTER_APP_NAME].status == "blocked",
+            timeout=SLOW_TIMEOUT,
+        )
+
+        logger.info("Relating mysql, mysqlrouter and application")
+        # Relate the database with mysqlrouter
+        await ops_test.model.relate(
+            f"{MYSQL_ROUTER_APP_NAME}:backend-database", f"{MYSQL_APP_NAME}:database"
+        )
+        # Relate mysqlrouter with data integrator
+        await ops_test.model.relate(
+            f"{DATA_INTEGRATOR}:mysql", f"{MYSQL_ROUTER_APP_NAME}:database"
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[MYSQL_ROUTER_APP_NAME], status="active", timeout=SLOW_TIMEOUT
+        )
+
+        # Relate mysqlrouter with application next
+        await ops_test.model.relate(
+            f"{APPLICATION_APP_NAME}:database", f"{MYSQL_ROUTER_APP_NAME}:database"
+        )
+
+        # Now, we should have one
+        await ops_test.model.wait_for_idle(
+            apps=[MYSQL_APP_NAME, MYSQL_ROUTER_APP_NAME, APPLICATION_APP_NAME, DATA_INTEGRATOR],
+            status="active",
+            raise_on_blocked=True,
+            timeout=SLOW_TIMEOUT,
+        )
+
+    # Ensure that the data inserted by sample application is present in the database
+    application_unit = application_app.units[0]
+    inserted_data = await get_inserted_data_by_application(application_unit)
+
+    mysql_unit = mysql_app.units[0]
+    mysql_unit_address = await get_unit_address(ops_test, mysql_unit.name)
+    server_config_credentials = await get_server_config_credentials(mysql_unit)
+
+    select_inserted_data_sql = [
+        f"SELECT data FROM continuous_writes_database.random_data WHERE data = '{inserted_data}'",
+    ]
+    selected_data = await execute_queries_on_unit(
+        mysql_unit_address,
+        server_config_credentials["username"],
+        server_config_credentials["password"],
+        select_inserted_data_sql,
+    )
+
+    assert len(selected_data) > 0
+    assert inserted_data == selected_data[0]
+
+    # Ensure the endpoints are set respectively to NodePort and ClusterIP for the data-integrator
+    # and the application-app
+    for app_name in [DATA_INTEGRATOR, APPLICATION_APP_NAME]:
+        try:
+            endpoint = yaml.safe_load(
+                subprocess.check_output(
+                    [
+                        "juju",
+                        "show-unit",
+                        f"{app_name}/0",
+                    ]
+                )
+            )[f"{app_name}/0"]["relation-info"][1]["application-data"]["endpoints"]
```

Comment:
The relation index `1` can possibly blow up here (e.g. if data-integrator/test-app adds a new peer relation).